core.go

```diff
@@ -96,35 +96,46 @@ func Strict(laxValidator validator.Validator) validator.Validator {
 	}
 }
 
+var strType = reflect.TypeOf("")
+var ifaceType = reflect.TypeOf(interface{}("a"))
+
 func compile(in interface{}) (validator.Validator, error) {
 	switch in.(type) {
-	case map[string]interface{}:
-		return compileMap(in.(map[string]interface{}))
-	case []interface{}:
-		return compileSlice(in.([]interface{}))
 	case isdef.IsDef:
 		return compileIsDef(in.(isdef.IsDef))
 	case nil:
 		// nil can't be handled by the default case of IsEqual
 		return compileIsDef(isdef.IsNil)
 	default:
-		// By default we just check reflection equality
-		return compileIsDef(isdef.IsEqual(in))
+		inVal := reflect.ValueOf(in)
+		switch inVal.Kind() {
+		case reflect.Map:
+			return compileMap(inVal)
+		case reflect.Slice:
+			return compileSlice(inVal)
+
+		default:
+			// By default we just check reflection equality
+			if inVal.Kind() == reflect.Map {
```

Comment:
This looks like it can be dropped because it's included as a case in the switch.